include/hipSYCL/glue/reflection.hpp

```diff
@@ -29,9 +29,11 @@
 #define HIPSYCL_REFLECTION_HPP
 
 template<class StructT>
-void __hipsycl_introspect_flattened_struct(int **num_flattened_members,
-                                      int **member_offsets, int **member_sizes,
-                                      int **member_kinds);
+void __hipsycl_introspect_flattened_struct(StructT *s,
```

Comment:
I've now added a pointer to StructT here and pass a pointer to the `s_copy` when calling the builtin. This seems to work properly. I first tried to make `s` here just of type StructT (no pointer) and pass `s` below directly. Then `getOperand(0)` on the builtin CallInst is a GetElementPtrInst but I couldn't get the type of the StructT this way, not sure what I did wrong. 

Also for some reason clang complains that the builtin is used but not defined as soon as you add the `StructT *` argument (I don't understand 100% why, it seems that for some stdpar calls the type `StructT` is a lambda and certain lambda types cannot  be shared across translation unit boundaries, see e.g. here https://stackoverflow.com/questions/76582942/lambda-variable-shared-between-translation-units). Therefore I added the empty definition `{}`.

Reply:
I think it can be even more simplified if we pass the `StructT` parameter by value to the builtin because we can then simply use something like
```cpp
llvm::Type *T = nullptr;
if (F->hasParamAttribute(0, llvm::Attribute::ByVal))
  T = F->getParamAttribute(0, llvm::Attribute::ByVal).getValueAsType();
else
  T = F->getArg(0)->getType();
StructTypeInfoForBuiltin[F] = getTypeInformation(T, M);
```
This would still have the problem that the builtin has to be defined and not just declared (due to the issue with lambdas mentioned above). Although we could probably add `__attribute__((optnone))` to ensure that it's not optimized away.

Reply:
> Edit: I guess this might not work since the call to the builtin could just be optimized away?

Yes.

> I think it can be even more simplified if we pass the StructT parameter by value 

I've experimented a lot with passing things by value. In my experience, it is inherently unreliable because there's not guarantee how it gets passed, since that depends on ABI. For example, on x86, if the struct is smaller than 16 bytes, the struct might be decomposed into its individual members and those might be passed as separate arguments. If it is passed as one argument, it may or may not be passed as pointer with LLVM byval attribute depending on platform etc.